File: app/src/main/java/ch/bailu/aat/providers/GpxProvider.java
package ch.bailu.aat.providers;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.OpenableColumns;
import android.support.annotation.NonNull;

import java.io.File;
import java.io.FileNotFoundException;


public class GpxProvider extends ContentProvider {
    private static final UnsupportedOperationException UNSUPORTED=
            new UnsupportedOperationException("Not supported by this provider");
    

    @Override
    public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {

        final String path = uri.getPath();

        if (path != null) {
            final File file = new File(uri.getPath());

            if (file.exists()) {
                return (ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY));
            }
        }
        throw UNSUPORTED;
    }

    
    @Override
    public boolean onCreate() {
        
        
        return true;
    }

    @Override
    public Cursor query(@NonNull Uri uri, String[] strings, String s, String[] strings2, String s2) {
        
        MatrixCursor cursor = new MatrixCursor(new String[]{OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE});
        
        
        final File file =  new File(uri.getPath());

        cursor.addRow(new Object[]{file.getName(), file.length()});
        return cursor;
    }

    @Override
    public String getType(@NonNull Uri uri) {
        final File file =  new File(uri.getPath());
        return mimeTypeFromFileName(file.getName());
    }
    
    public static String mimeTypeFromFileName(String name) {
        if (name.endsWith(".gpx")) return "application/gpx+xml";
        else if (name.endsWith(".osm")) return "application/xml";
        return null;
    }

    @Override
    public Uri insert(@NonNull Uri uri, ContentValues contentValues) {
        throw UNSUPORTED;
    }

    @Override
    public int delete(@NonNull Uri uri, String s, String[] strings) {
        throw UNSUPORTED;
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues contentValues, String s, String[] strings) {
        throw UNSUPORTED;
    }
    
}

File: aat-android/src/main/java/ch/bailu/aat/providers/GpxProvider.kt
package ch.bailu.aat.providers

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.provider.OpenableColumns
import java.io.File
import java.io.FileNotFoundException

class GpxProvider : ContentProvider() {
    @Throws(FileNotFoundException::class)
    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor? {
        val path = uri.path
        if (path != null) {
            val file = File(path)
            if (file.exists()) {
                return ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
            }
        }
        throw UNSUPPORTED
    }

    override fun onCreate(): Boolean {
        return true
    }

    override fun query(
        uri: Uri,
        strings: Array<String>?,
        s: String?,
        strings2: Array<String>?,
        s2: String?
    ): Cursor {
        val cursor = MatrixCursor(arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE))
        val path = uri.path

        if (path != null) {
            val file = File(path)
            cursor.addRow(arrayOf<Any>(file.name, file.length()))
        }
        return cursor
    }

    override fun getType(uri: Uri): String? {
        val path = uri.path
        return if (path != null) {
            mimeTypeFromFileName(File(path).name)
        } else {
            null
        }
    }

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? {
        throw UNSUPPORTED
    }

    override fun delete(uri: Uri, s: String?, strings: Array<String>?): Int {
        throw UNSUPPORTED
    }

    override fun update(
        uri: Uri,
        contentValues: ContentValues?,
        s: String?,
        strings: Array<String>?
    ): Int {
        throw UNSUPPORTED
    }

    companion object {
        private val UNSUPPORTED = UnsupportedOperationException("Not supported by this provider")
        fun mimeTypeFromFileName(name: String): String? {
            if (name.endsWith(".gpx")) return "application/gpx+xml" else if (name.endsWith(".osm")) return "application/xml"
            return null
        }
    }
}
